src/librbd/operation/DisableFeaturesRequest.cc

```diff
@@ -0,0 +1,599 @@
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *DisableFeaturesRequest<I>::handle_get_mirror_mode(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << ": r=" << *result << dendl;
+
+  if (*result == 0) {
+    bufferlist::iterator it = m_out_bl.begin();
+    *result = cls_client::mirror_mode_get_finish(&it, &m_mirror_mode);
+  }
+
+  if (*result < 0 && *result != -ENOENT) {
+    lderr(cct) << "failed to retrieve pool mirror mode: "
+               << cpp_strerror(*result) << dendl;
+    return handle_finish(*result);
+  }
+
+  ldout(cct, 20) << this << " " << __func__ << ": m_mirror_mode="
+                 << m_mirror_mode << dendl;
+
+  send_get_mirror_image();
+  return nullptr;
+}
+
+template <typename I>
+void DisableFeaturesRequest<I>::send_get_mirror_image() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  if (m_mirror_mode != cls::rbd::MIRROR_MODE_IMAGE) {
+    send_disable_mirror_image();
+    return;
+  }
+
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_image_get_start(&op, image_ctx.id);
+
+  using klass = DisableFeaturesRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_get_mirror_image>(this);
+  m_out_bl.clear();
+  int r = image_ctx.md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *DisableFeaturesRequest<I>::handle_get_mirror_image(int *result) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  cls::rbd::MirrorImage mirror_image;
+
+  if (*result == 0) {
+    bufferlist::iterator it = m_out_bl.begin();
+    *result = cls_client::mirror_image_get_finish(&it, &mirror_image);
+  }
+
+  if (*result < 0 && *result != -ENOENT) {
+    lderr(cct) << "failed to retrieve pool mirror image: "
+               << cpp_strerror(*result) << dendl;
+    return handle_finish(*result);
+  }
+
+  if (mirror_image.state == cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+    lderr(cct) << "cannot disable journaling: image mirroring "
+               << "enabled and mirror pool mode set to image"
+               << dendl;
+    *result = -EINVAL;
+    return handle_finish(*result);
+  }
+
+  send_disable_mirror_image();
+  return nullptr;
+}
+
+template <typename I>
+void DisableFeaturesRequest<I>::send_disable_mirror_image() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  ldout(cct, 20) << this << " " << __func__ << dendl;
+
+  Context *ctx = create_context_callback<
+    DisableFeaturesRequest<I>,
+    &DisableFeaturesRequest<I>::handle_disable_mirror_image>(this);
+
+  mirror::DisableRequest<I> *req =
+    mirror::DisableRequest<I>::create(&image_ctx, false, true, ctx);
```

Comment:
I am ok with this change. I think you can do it yourself as an additional commit?

@dillaman What do you think?